File: push-be/internal/domain/usecase/billing.go
package usecase

import (
	"context"
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"strconv"
	"strings"
	"time"

	"github.com/google/uuid"

	"push-be/internal/domain"
)

const stripeWebhookTolerance = 5 * time.Minute

type BillingConfig struct {
	Configured    bool
	Prices        map[string]string
	SuccessURL    string
	CancelURL     string
	ReturnURL     string
	WebhookSecret string
}

type BillingService struct {
	billing domain.BillingStore
	stripe  domain.StripeGateway
	uow     domain.UnitOfWork
	cfg     BillingConfig
}

func NewBillingService(billing domain.BillingStore, stripe domain.StripeGateway, uow domain.UnitOfWork, cfg BillingConfig) *BillingService {
	return &BillingService{billing: billing, stripe: stripe, uow: uow, cfg: cfg}
}

func (s *BillingService) Status(ctx context.Context, u *domain.User) (*domain.BillingStatus, error) {
	bal, err := s.billing.LastBalance(ctx, u.ID)
	if err != nil {
		return nil, domain.Internal()
	}
	st := &domain.BillingStatus{
		SubscriptionStatus: u.SubscriptionStatus,
		PeriodEndsAt:       u.PeriodEndsAt,
		BalanceMicroCredits: bal,
	}
	if st.SubscriptionStatus == "" {
		st.SubscriptionStatus = domain.SubStatusNone
	}
	if u.Plan != "" {
		plan := u.Plan
		st.Plan = &plan
	}
	return st, nil
}

func (s *BillingService) Checkout(ctx context.Context, u *domain.User, planID string) (*domain.StripeCheckout, error) {
	if !s.cfg.Configured {
		return nil, domain.NotConfigured("billing is not configured")
	}
	if planID == domain.PlanFree || !domain.ValidPlan(planID) {
		return nil, domain.ValidationField("planId", "unsupported plan")
	}
	price := s.cfg.Prices[planID]
	if price == "" {
		return nil, domain.NotConfigured("plan price is not configured")
	}
	customerID := ""
	if u.StripeCustomerID != nil {
		customerID = *u.StripeCustomerID
	}
	sess, err := s.stripe.CreateCheckoutSession(ctx, price, customerID, u.ID.String(), planID, s.cfg.SuccessURL, s.cfg.CancelURL)
	if err != nil {
		return nil, domain.ProviderError("stripe checkout session failed")
	}
	return sess, nil
}

func (s *BillingService) Portal(ctx context.Context, u *domain.User) (string, error) {
	if !s.cfg.Configured {
		return "", domain.NotConfigured("billing is not configured")
	}
	if u.StripeCustomerID == nil || *u.StripeCustomerID == "" {
		return "", domain.InvalidTransition("no billing account")
	}
	url, err := s.stripe.CreatePortalSession(ctx, *u.StripeCustomerID, s.cfg.ReturnURL)
	if err != nil {
		return "", domain.ProviderError("stripe portal session failed")
	}
	return url, nil
}

func (s *BillingService) Ledger(ctx context.Context, userID uuid.UUID, page domain.PageRequest) (domain.Page[domain.LedgerEntry], error) {
	limit := page.EffectiveLimit()
	items, err := s.billing.ListLedger(ctx, userID, page, limit+1)
	if err != nil {
		return domain.Page[domain.LedgerEntry]{}, domain.Internal()
	}
	return domain.NewPage(items, limit, func(e domain.LedgerEntry) domain.Cursor {
		return domain.Cursor{CreatedAt: e.CreatedAt, ID: e.ID}
	}), nil
}

func VerifyStripeSignature(payload []byte, header, secret string, now time.Time) error {
	var ts int64
	sigs := []string{}
	for _, part := range strings.Split(header, ",") {
		kv := strings.SplitN(strings.TrimSpace(part), "=", 2)
		if len(kv) != 2 {
			continue
		}
		switch kv[0] {
		case "t":
			v, err := strconv.ParseInt(kv[1], 10, 64)
			if err != nil {
				return errors.New("bad timestamp")
			}
			ts = v
		case "v1":
			sigs = append(sigs, kv[1])
		}
	}
	if ts == 0 || len(sigs) == 0 {
		return errors.New("missing signature")
	}
	if d := now.Unix() - ts; d > int64(stripeWebhookTolerance.Seconds()) || d < -int64(stripeWebhookTolerance.Seconds()) {
		return errors.New("stale timestamp")
	}
	mac := hmac.New(sha256.New, []byte(secret))
	fmt.Fprintf(mac, "%d.", ts)
	mac.Write(payload)
	expected := hex.EncodeToString(mac.Sum(nil))
	for _, sig := range sigs {
		if hmac.Equal([]byte(sig), []byte(expected)) {
			return nil
		}
	}
	return errors.New("signature mismatch")
}

func (s *BillingService) HandleWebhook(ctx context.Context, payload []byte, sigHeader string) error {
	if !s.cfg.Configured || s.cfg.WebhookSecret == "" {
		return domain.NotConfigured("billing is not configured")
	}
	if err := VerifyStripeSignature(payload, sigHeader, s.cfg.WebhookSecret, time.Now()); err != nil {
		return domain.Validation("invalid webhook signature")
	}
	var event struct {
		ID   string `json:"id"`
		Type string `json:"type"`
		Data struct {
			Object json.RawMessage `json:"object"`
		} `json:"data"`
	}
	if err := json.Unmarshal(payload, &event); err != nil || event.ID == "" {
		return domain.Validation("invalid webhook payload")
	}
	fresh, err := s.billing.RecordStripeEvent(ctx, event.ID, event.Type, time.Now().UTC())
	if err != nil {
		return domain.Internal()
	}
	if !fresh {
		return nil
	}
	switch event.Type {
	case "checkout.session.completed":
		return s.handleCheckoutCompleted(ctx, event.Data.Object)
	case "customer.subscription.updated", "customer.subscription.deleted":
		return s.handleSubscriptionChanged(ctx, event.Type, event.Data.Object)
	default:
		return nil
	}
}

func (s *BillingService) handleCheckoutCompleted(ctx context.Context, raw json.RawMessage) error {
	var obj struct {
		ID                string `json:"id"`
		ClientReferenceID string `json:"client_reference_id"`
		Customer          string `json:"customer"`
		Metadata          struct {
			PlanID string `json:"planId"`
		} `json:"metadata"`
	}
	if err := json.Unmarshal(raw, &obj); err != nil {
		return domain.Validation("invalid session payload")
	}
	userID, err := uuid.Parse(obj.ClientReferenceID)
	if err != nil {
		return domain.Validation("invalid session reference")
	}
	planID := obj.Metadata.PlanID
	if !domain.ValidPlan(planID) || planID == domain.PlanFree {
		return domain.Validation("unsupported plan")
	}
	credits := domain.PlanCreditsMicroCredits[planID]
	return s.uow.Do(ctx, func(ctx context.Context) error {
		if err := s.billing.UpdateSubscription(ctx, userID, planID, domain.SubStatusActive, &obj.Customer, nil); err != nil {
			return err
		}
		bal, err := s.billing.LastBalance(ctx, userID)
		if err != nil {
			return err
		}
		return s.billing.AppendLedger(ctx, &domain.LedgerEntry{
			ID: uuid.New(), UserID: userID, Type: domain.LedgerPurchase,
			AmountMicroCredits: credits, BalanceAfter: bal + credits,
			ReferenceID: &obj.ID, CreatedAt: time.Now().UTC(),
		})
	})
}

func (s *BillingService) handleSubscriptionChanged(ctx context.Context, typ string, raw json.RawMessage) error {
	var obj struct {
		Customer           string `json:"customer"`
		Status             string `json:"status"`
		CurrentPeriodEnd   int64  `json:"current_period_end"`
	}
	if err := json.Unmarshal(raw, &obj); err != nil || obj.Customer == "" {
		return domain.Validation("invalid subscription payload")
	}
	var periodEnd *time.Time
	if obj.CurrentPeriodEnd > 0 {
		t := time.Unix(obj.CurrentPeriodEnd, 0).UTC()
		periodEnd = &t
	}
	status := domain.SubStatusActive
	if typ == "customer.subscription.deleted" || obj.Status == "canceled" {
		status = domain.SubStatusCanceled
	}
	if err := s.billing.UpdateSubscriptionByCustomer(ctx, obj.Customer, status, periodEnd); err != nil {
		return domain.Internal()
	}
	return nil
}

File: push-be/app/domain/services/billing.py
from __future__ import annotations
import hashlib
import hmac as hmac_mod
import json
from datetime import datetime, timezone
from uuid import UUID, uuid4

from app.db import DB
from app.domain import entities as ent
from app.domain.errors import (
    internal, invalid_transition, not_configured, provider_error,
    validation, validation_field,
)
from app.infrastructure.store_integrations import BillingStore

STRIPE_WEBHOOK_TOLERANCE = 300


def _now() -> datetime:
    return datetime.now(timezone.utc)


def verify_stripe_signature(payload: bytes, header: str, secret: str,
                            now: datetime) -> None:
    ts = 0
    sigs = []
    for part in header.split(","):
        kv = part.strip().split("=", 1)
        if len(kv) != 2:
            continue
        if kv[0] == "t":
            try:
                ts = int(kv[1])
            except ValueError:
                raise ValueError("bad timestamp")
        elif kv[0] == "v1":
            sigs.append(kv[1])
    if not ts or not sigs:
        raise ValueError("missing signature")
    d = int(now.timestamp()) - ts
    if abs(d) > STRIPE_WEBHOOK_TOLERANCE:
        raise ValueError("stale timestamp")
    expected = hmac_mod.new(
        secret.encode(), f"{ts}.".encode() + payload, hashlib.sha256).hexdigest()
    if not any(hmac_mod.compare_digest(sig, expected) for sig in sigs):
        raise ValueError("signature mismatch")


class BillingService:
    def __init__(self, db: DB, stripe, configured: bool, prices: dict,
                 success_url: str, cancel_url: str, return_url: str,
                 webhook_secret: str):
        self.db = db
        self.billing = BillingStore(db)
        self.stripe = stripe
        self.configured = configured
        self.prices = prices
        self.success_url = success_url
        self.cancel_url = cancel_url
        self.return_url = return_url
        self.webhook_secret = webhook_secret

    async def status(self, u: ent.User) -> ent.BillingStatus:
        try:
            bal = await self.billing.last_balance(u.id)
        except Exception:
            raise internal()
        st = ent.BillingStatus(
            subscription_status=u.subscription_status or ent.SUB_NONE,
            plan=u.plan or None, period_ends_at=u.period_ends_at,
            balance_micro_credits=bal)
        return st

    async def checkout(self, u: ent.User, plan_id: str) -> ent.StripeCheckout:
        if not self.configured:
            raise not_configured("billing is not configured")
        if plan_id == ent.PLAN_FREE or not ent.valid_plan(plan_id):
            raise validation_field("planId", "unsupported plan")
        price = self.prices.get(plan_id, "")
        if not price:
            raise not_configured("plan price is not configured")
        try:
            return await self.stripe.create_checkout_session(
                price, u.stripe_customer_id or "", str(u.id), plan_id,
                self.success_url, self.cancel_url)
        except Exception:
            raise provider_error("stripe checkout session failed")

    async def portal(self, u: ent.User) -> str:
        if not self.configured:
            raise not_configured("billing is not configured")
        if not u.stripe_customer_id:
            raise invalid_transition("no billing account")
        try:
            return await self.stripe.create_portal_session(
                u.stripe_customer_id, self.return_url)
        except Exception:
            raise provider_error("stripe portal session failed")

    async def ledger(self, user_id: UUID, page):
        try:
            return await self.billing.list_ledger(user_id, page)
        except Exception:
            raise internal()

    async def handle_webhook(self, payload: bytes, sig_header: str) -> None:
        if not self.configured or not self.webhook_secret:
            raise not_configured("billing is not configured")
        try:
            verify_stripe_signature(payload, sig_header, self.webhook_secret,
                                    datetime.now())
        except ValueError:
            raise validation("invalid webhook signature")
        try:
            event = json.loads(payload)
        except Exception:
            event = {}
        if not event.get("id"):
            raise validation("invalid webhook payload")
        try:
            fresh = await self.billing.record_stripe_event(
                event["id"], event.get("type", ""), _now())
        except Exception:
            raise internal()
        if not fresh:
            return
        obj = (event.get("data") or {}).get("object") or {}
        if event["type"] == "checkout.session.completed":
            await self._checkout_completed(obj)
        elif event["type"] in ("customer.subscription.updated",
                               "customer.subscription.deleted"):
            await self._subscription_changed(event["type"], obj)

    async def _checkout_completed(self, obj: dict) -> None:
        try:
            user_id = UUID(obj.get("client_reference_id") or "")
        except ValueError:
            raise validation("invalid session reference")
        plan_id = (obj.get("metadata") or {}).get("planId", "")
        if not ent.valid_plan(plan_id) or plan_id == ent.PLAN_FREE:
            raise validation("unsupported plan")
        credits = ent.PLAN_CREDITS_MICRO[plan_id]
        session_id = obj.get("id", "")
        customer = obj.get("customer", "")

        async def work():
            await self.billing.update_subscription(
                user_id, plan_id, ent.SUB_ACTIVE, customer, None)
            bal = await self.billing.last_balance(user_id)
            await self.billing.append_ledger(ent.LedgerEntry(
                id=uuid4(), user_id=user_id, type=ent.LEDGER_PURCHASE,
                amount_micro_credits=credits, balance_after=bal + credits,
                reference_id=session_id, created_at=_now()))

        await self.db.do(work)

    async def _subscription_changed(self, typ: str, obj: dict) -> None:
        customer = obj.get("customer", "")
        if not customer:
            raise validation("invalid subscription payload")
        period_end = None
        if obj.get("current_period_end"):
            period_end = datetime.fromtimestamp(
                obj["current_period_end"], timezone.utc)
        status = ent.SUB_ACTIVE
        if typ == "customer.subscription.deleted" or obj.get("status") == "canceled":
            status = ent.SUB_CANCELED
        try:
            await self.billing.update_subscription_by_customer(
                customer, status, period_end)
        except Exception:
            raise internal()
